### PR TITLE
Feature/farm 389 article repository

### PR DESCRIPTION
**Pull request by LeeMatthewHiggins**
[FARM-389 - Article Repo Plus much more]

#### 📲 What

This is a major refactor from redux to Bloc pattern. It´s only the Discover tab that will use it.
The back history is fixed in the context of that tab, but if you switch tabs the Discover tab will loose its state. This will be fixed in anther PR when we remove the redux pattern from other tabs. 

Refactor of service layer to model Flamelink - implement filters for production and dev, and locale.
Introduction of Repos (hide the Flamelink specifics)
Refactor of image system - will need further improvements as we go
Abstraction of image URL fetching (so we can give width, height context and choose the correct size)
This needs the actual function to make that choice implementing, atm just chooses the first size.
Introduction of unit tests (limited)  

#### 🤔 Why
		
This is needed to build faster and be able to test.
To fix app wide issues with state and navigation.
		
#### 🛠 How
		
       [Model]    ->               [Bloc]             -> [View]  
   [repo , model] -> [ViewModelProvider, Transformer] -> [viewModel, widget]

#### 👀 See
		
Screenshots / external resources
		 
#### ✅ Acceptance criteria

- [ ] Design Review for UI with BA completed. 
- [ ] Manually tested and verified on Android device/emulator (min. Lollipop 5.1)
- [ ] Showcase video / automation test
- [X] Passing all tests
- [X] Rebased/merged with latest changes from development and re-tested
- [ ] Removed unsed comments. TODOs must have JIRA for future work.

#### Coding Standards Checklist
- [X] meets coding standards

#### 🕵️‍♂️ How to test

Go to discover section and play :)
Look for the ticket number in the Debug section

#### Reviewers

@farmsmart/amido @farmsmart/wamf

### CODE REVIEW

**Review by LeeMatthewHiggins — COMMENTED** [review]
The PR is too big. I did try and break it up, but had a lot of cascading dependencies in my work. If we get this pain over PR´s should be small in future. Sorry.

File: lib/data/repositories/articles_directory_repository.dart [hunk 1]
```diff
@@ -5,6 +5,8 @@ import 'package:farmsmart_flutter/data/model/article_entity.dart';
 import 'package:farmsmart_flutter/data/model/stage_entity.dart';
 import 'package:flutter/foundation.dart';
 
+// LH this file is deprecated !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Comment (davidfranquet) [inline lib/data/repositories/articles_directory_repository.dart]: Please add the deprecated annotation: `@Deprecated('message')`

File: lib/data/bloc/Transformer.dart [hunk 2]
```diff
@@ -0,0 +1,16 @@
+//typedef Transformer<A,B> = B Function(A);
+import 'package:flutter/foundation.dart';
+
+abstract class ObjectTransformer<A,B> {
+    B transform({@required A from});
+}
+
+//Helper functions for safe casting when transforming 
+T castOrPlaceholder<T>(dynamic from, T placeholder) {
```
Comment (davidfranquet) [inline lib/data/bloc/Transformer.dart:9]: 👍

File: lib/data/repositories/implementation/ArticlesRepositoryFlamelink.dart [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  article.related = ArticleEntityCollectionFlamelink(collection: articleCollection);
+  article.images = ImageEntityCollectionFlamelink(collection: imageCollection);
+  return article;
+}
+
+class ArticleEntityCollectionFlamelink implements EntityCollection<ArticleEntity> {
+  final FlamelinkDocumentCollection _collection;
+
+  ArticleEntityCollectionFlamelink({FlamelinkDocumentCollection collection})
+      :_collection = collection;
```
Comment (davidfranquet) [inline lib/data/repositories/implementation/ArticlesRepositoryFlamelink.dart]:
```suggestion
      : _collection = collection;
```

File: lib/redux/app/app_reducer.dart [hunk 4]
```diff
@@ -1,11 +1,9 @@
 import 'package:farmsmart_flutter/redux/app/app_state.dart';
 import 'package:farmsmart_flutter/redux/home/home_reducer.dart';
 import 'package:farmsmart_flutter/redux/home/myPlot/my_plot_reducer.dart';
-import 'package:farmsmart_flutter/redux/home/discover/discover_reducer.dart';
 
 AppState appReducer(AppState state, dynamic action) =>
     AppState(
         homeState: homeReducer(state.homeState ,action),
-        myPlotState: myPlotReducer(state.myPlotState, action),
-        discoverState: discoverReducer(state.discoverState, action)
```
Comment (davidfranquet) [inline lib/redux/app/app_reducer.dart:10]: 👏

File: lib/ui/common/ContextualAppBar.dart [hunk 5]
```diff
@@ -0,0 +1,62 @@
+
+import 'package:flutter/cupertino.dart';
```
Comment (davidfranquet) [inline lib/ui/common/ContextualAppBar.dart]: Do we need this import? 🤔

File: lib/ui/myplot/my_plot_current_stage_screen.dart [hunk 6]
```diff
@@ -97,7 +97,7 @@ Widget buildListOfRelatedArticles(ArticleEntity articleData, MyPlotViewModel vie
 
 _buildListItemImage(ArticleEntity articleData) {
   return NetworkImageFromFuture(
-      articleData.imageUrl,
+      null,
```
Comment (davidfranquet) [inline lib/ui/myplot/my_plot_current_stage_screen.dart:100]: Mistake?

File: lib/utils/MockString.dart [hunk 7]
```diff
@@ -0,0 +1,39 @@
+import 'dart:math';
+
+class MockString {
```
Comment (davidfranquet) [inline lib/utils/MockString.dart:3]: 👍

**Comment by davidfranquet** [comment]
Good job @LeeMatthewHiggins ! I really like this architecture 👏

File: lib/data/repositories/articles_directory_repository.dart (hunk 1 above)
Comment (LeeMatthewHiggins) [reply lib/data/repositories/articles_directory_repository.dart]: nice, didn't know how to do that. 👍

File: lib/ui/myplot/my_plot_current_stage_screen.dart (hunk 6 above)
Comment (LeeMatthewHiggins) [reply lib/ui/myplot/my_plot_current_stage_screen.dart:100]: I had to remove that to be able to compile, as I removed all the imageURL mess from the Article entities, now uses the imageProvider interface, I will restore this in the MyPlot work. I´ll add a TODO:

File: lib/data/repositories/implementation/ArticlesRepositoryFlamelink.dart (hunk 3 above)
Comment (LeeMatthewHiggins) [reply lib/data/repositories/implementation/ArticlesRepositoryFlamelink.dart]: 👍

File: lib/ui/common/ContextualAppBar.dart (hunk 5 above)
Comment (LeeMatthewHiggins) [reply lib/ui/common/ContextualAppBar.dart]: nope, just being an apple fanboy.

Comment (LeeMatthewHiggins) [reply lib/ui/common/ContextualAppBar.dart]: removed

File: lib/data/bloc/article/ArticleDetailTransformer_test.dart [hunk 8]
```diff
@@ -0,0 +1,26 @@
+import 'package:farmsmart_flutter/data/bloc/article/ArticleDetailTransformer.dart';
+import 'package:farmsmart_flutter/data/bloc/article/ArticleListItemViewModelTransformer.dart';
+import 'package:farmsmart_flutter/data/repositories/article/implementation/MockArticlesRepository.dart';
+import 'package:intl/intl.dart';
+import 'package:test/test.dart';
+
+final testArticle = MockArticle.build();
+
+void main() {
+  test('Article detail viewmodel transfomed from article correctly', () {
```
Comment (davidfranquet) [inline lib/data/bloc/article/ArticleDetailTransformer_test.dart]: Tests should go on root/test/ folder

**Review by raubreysmith — DISMISSED** [review]
As discussed there are several issues raised due to the scope of work however these will be covered by other WiP or separate bugs

**Review by isaacmartinezimv — APPROVED** [review]

**Pull request merged**